exporter/prometheusexporter/collector.go

```diff
@@ -403,4 +428,50 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
 		ch <- m
 		c.logger.Debug(fmt.Sprintf("metric served: %s", m.Desc().String()))
 	}
+	c.cleanupMetricFamilies()
+}
+
+func (c *collector) validateMetrics(name, description string, metricType *dto.MetricType) (help string, err error) {
+	now := time.Now()
+	v, exist := c.metricFamilies.Load(name)
+	if !exist {
+		c.metricFamilies.Store(name, metricFamily{
+			lastSeen: now,
+			mf: &dto.MetricFamily{
+				Name: proto.String(name),
+				Help: proto.String(description),
+				Type: metricType,
+			},
+		})
+		return description, nil
+	}
+	emf := v.(metricFamily)
+	if emf.mf.GetType() != *metricType {
+		return "", fmt.Errorf("instrument type conflict, using existing type definition. instrument: %s, existing: %s, dropped: %s", name, emf.mf.GetType(), *metricType)
+	}
+	emf.lastSeen = now
+	c.metricFamilies.Store(name, emf)
+	if emf.mf.GetHelp() != description {
+		c.logger.Info(
+			"Instrument description conflict, using existing",
+			zap.String("instrument", name),
+			zap.String("existing", emf.mf.GetHelp()),
+			zap.String("dropped", description),
+		)
+	}
+	return emf.mf.GetHelp(), nil
+}
+
+func (c *collector) cleanupMetricFamilies() {
+	expirationTime := time.Now().Add(-c.metricExpiration)
+
+	c.metricFamilies.Range(func(key, value any) bool {
+		v := value.(metricFamily)
+		if expirationTime.After(v.lastSeen) {
+			c.logger.Debug(fmt.Sprintf("metric expired: %s", key))
```

Comment:
any way to avoid calling fmt.Sprintf here

Reply:
thanks @kristinapathak